sor/src/main/java/com/bazaarvoice/emodb/sor/audit/s3/AthenaAuditWriter.java

```diff
@@ -0,0 +1,536 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.*;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Audit writer implementation which writes all audits as GZIP'd jsonl documents to S3 partitioned by date.
+ * This format allows audit queries to be carried out by Athena, Amazon's Presto implementation over S3 documents.
+ *
+ * This audit writer favors fast, non-blocking calls to {@link #persist(String, String, Audit, long)} over guaranteeing
+ * a completely loss-less audit history.  To achieve this all audits are written to an in-memory queue.  That queue
+ * is written to a local log file until it has reached a maximum size or age, both configurable in the constructor.
+ * At this time the file is asynchronously GZIP'd then delivered to S3.  Once the file is delivered it is deleted from
+ * the local host.
+ *
+ * Each stage has multiple layers of recovery, ensuring that once a line is written to a file that file will eventually
+ * be delivered to S3.  The exceptions to this which can cause audit loss are:
+ *
+ * <ol>
+ *     <li>The process is terminated while unwritten audits are still in the audit queue.</li>
+ *     <li>The host itself terminates before all files are delivered to S3.</li>
+ * </ol>
+ */
+public class AthenaAuditWriter implements AuditStore {
+
+    private final static Logger _log = LoggerFactory.getLogger(AthenaAuditWriter.class);
+
+    private final static DateTimeFormatter LOG_FILE_DATE_FORMATTER =
+            DateTimeFormatter.ofPattern("yyyyMMddHHmmss").withZone(ZoneOffset.UTC);
+
+    private final static String OPEN_FILE_SUFFIX = ".log.tmp";
+    private final static String CLOSED_FILE_SUFFIX = ".log";
+    private final static String COMPRESSED_FILE_SUFFIX = ".log.gz";
+
+    private final static long DEFAULT_MAX_FILE_SIZE = Size.megabytes(10).toBytes();
+    private final static Duration DEFAULT_MAX_BATCH_TIME = Duration.ofMinutes(2);
+
+    private final AmazonS3 _s3;
+    private final String _s3Bucket;
+    private final String _s3AuditRoot;
+    private final long _maxFileSize;
+    private final long _maxBatchTimeMs;
+    private final File _stagingDir;
+    private final String _logFilePrefix;
+    private final BlockingQueue<QueuedAudit> _auditQueue;
+    private final Clock _clock;
+    private final ObjectWriter _objectWriter;
+    private final ConcurrentMap<Long, AuditOutput> _openAuditOutputs = Maps.newConcurrentMap();
+
+    private ScheduledExecutorService _auditService;
+    private ExecutorService _fileTransferService;
+    private AuditOutput _mruAuditOutput;
+    private boolean _fileTransfersEnabled;
+    private final RateLimitedLog _rateLimitedLog;
+
+    @Inject
+    public AthenaAuditWriter(AuditWriterConfiguration config, ObjectMapper objectMapper, Clock clock,
```

Comment:
Normally injecting application config into a lower-level object would set off red flags.  However, in this case the config is really tailored for this object, and you've addressed the main concern -- losing control over unit testing by constructing instead of injecting all dependencies, particularly in this case `AmazonS3` -- by providing a secondary constructor.  So I'll put out there whether you agree and would like to refactor injection, but it's fine as-is.

Reply:
Yeah I grappled with this issue for a while, and ultimately came to the conclusion that because of the all of these config values apply solely to `AthenaAuditWriter`, that is was best to let all of the construction just occur in the class itself, rather than spread this logic all over our codebase in a way that makes it hard to track.